### ISSUE TITLE
jax.grad precision: float32 gradients of bfloat16 weights

### ISSUE THREAD

**Issue by martin-marek** [jax-ml/jax#30337]
I am interested in fine-tuning a large model in a memory-constrained setting. I would like to store model weights in `bf16` and compute activations and gradients in `fp32`. However, I cannot figure out how to compute gradients in a different precision from the weights.

My understanding is that by default, `jax.grad` uses the same precision for gradients as the weights. Is there any way to get around this?

**Comment by rdyro** [comment]
Can you try, under the same jit, computing the gradients wrt parameters you first cast to the higher dtype? I believe the compiler should make the correct decision of only upcasting before the parameter is needed.

Something like:
```python
@jax.jit
def grad(params, *args):
  params = jax.tree.map(lambda x: x.astype(jnp.float32), params)
  return jax.grad(loss_fn)(params, args)
```

**Comment by martin-marek** [comment]
Thank you @rdyro for your quick response!

Indeed your solution works perfectly, it's amazing that the JIT compiler just handles this correctly 🙂
